services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateManager.java

```diff
@@ -113,221 +213,453 @@ public boolean shouldSkipClass(Class<?> clazz) {
+                portStates,
+                3);
+      }
+    }
 
     // construct sb msg by ip
-    Map<String, Goalstate.GoalState> goalStateHashMap = new HashMap<>();
+    Map<String, Goalstate.GoalState> goalStateConcurrentHashMap = new ConcurrentHashMap<>();
     // TODO would opt this part when perf needed
-    mapGroupedByHostIp.entrySet().stream()
+    hostIpSubnetIdsMap
+        .keySet()
         .forEach(
-            eachGSOnSingleIP -> {
+            currentGroupHostIp -> {
               Set<PortState> portStateHashSet = new HashSet<>();
+              Map<String, Neighbor.NeighborState> neighborStates = new ConcurrentHashMap<>();
               Set<Subnet.SubnetState> subnetStateSet = new HashSet();
               Set<Vpc.VpcState> vpcStateSet = new HashSet();
-                List<DHCP.DHCPState> dhcpStateList = new ArrayList();
-
-                final List<InternalPortEntity> internalPortEntitySet = eachGSOnSingleIP.getValue();
-              boolean m = false;
-              internalPortEntitySet.stream()
-                  .forEach(
-                      portStateWithEverythingFilledNB -> {
-                        Set<Port.PortConfiguration.HostInfo> neighborSB = new HashSet();
-                        if (portStateWithEverythingFilledNB.getInternalNeighborInfo1() != null) {
-                          neighborSB = new HashSet();
-                          for (NeighborInfo neighborInfo :
-                              portStateWithEverythingFilledNB.getInternalNeighborInfo1()) {
-                            Port.PortConfiguration.HostInfo build =
-                                Port.PortConfiguration.HostInfo.newBuilder()
-                                    .setIpAddress(neighborInfo.getHostIp())
-                                    .setMacAddress(neighborInfo.getPortId())
-                                    .build();
-                            neighborSB.add(build);
+              Set<DHCP.DHCPState> dhcpStateList = new HashSet();
+              final List<InternalPortEntity> internalPortEntitySet =
+                  mapGroupedByHostIp.get(currentGroupHostIp);
+              if (internalPortEntitySet != null) {
+                internalPortEntitySet.forEach(
+                    portStateWithEverythingFilledNB -> {
+                      List<FixedIp> fixedIps = new ArrayList();
+                      boolean isExistingPort = false;
+                      final List<PortEntity.FixedIp> fixedIps1 =
+                          portStateWithEverythingFilledNB.getFixedIps();
+                      for (PortEntity.FixedIp fixedIp : fixedIps1) {
+                        fixedIps.add(
+                            FixedIp.newBuilder()
+                                .setSubnetId(fixedIp.getSubnetId())
+                                .setIpAddress(fixedIp.getIpAddress())
+                                .build());
+                        for (NeighborInfo internalPortEntity : portIdNeighborInfoMap.values()) {
```

Comment:
This block of codes have four for-loop, potentially quite costly. Pls try to improve this part.

Reply:
removed one loop in 7ccbc91 , for extra check we need to optimized the contract